engine/configurator/conf_cyaml.c

```diff
@@ -0,0 +1,796 @@
+ */
+const cyaml_schema_value_t history_entry_schema = {
+
+        CYAML_VALUE_MAPPING(CYAML_FLAG_DEFAULT,
+                            history_entry,
+                            history_entry_fields_schema),
+};
+
+/* The history field definitions schema is an array.
+ *
+ * All the field entries will refer to their parent mapping's structure,
+ * in this case, 'history_seq'.
+ */
+const cyaml_schema_field_t history_fields_schema[] = {
+    /* The only field is the register, unregister, add or something.
+     *
+     * YAML key: "history".
+     * C structure member for this key: "entries".
+     *
+     * Its CYAML type is a sequence.
+     *
+     * Since it's a sequence type, the value schema for its entries must
+     * be provided too. In this case, it's history_entry_schema.
+     *
+     * Since it's not a sequence of a fixed-length, we must tell CYAML
+     * where the sequence entry count is to be stored. In this case, it
+     * goes in the "entries_count" C structure member in
+     * history_seq.
+     * Since this is "entries" with the "_count" postfix, we can use
+     * the following macro, which assumes a postfix of "_count" in the
+     * struct member name.
+     */
+    CYAML_FIELD_SEQUENCE(
+                         "history",
+                         CYAML_FLAG_POINTER | CYAML_FLAG_OPTIONAL,
+                         history_seq, entries,
+                         &history_entry_schema, 0, CYAML_UNLIMITED),
+    /* The field array must be terminated by an entry with a NULL key.
+     * Here we use the CYAML_FIELD_END macro for that. */
+    CYAML_FIELD_END
+};
+
+/* Top-level schema of the history. The top level value for the history is
+ * a sequence.
+ *
+ * Its fields are defined in history_entry_fields_schema.
+ */
+const cyaml_schema_value_t history_schema = {
+    CYAML_VALUE_MAPPING(CYAML_FLAG_POINTER,
+                        history_seq, history_fields_schema),
+};
+
+
+/******************************************************************************
+ * Actual code to load and save YAML doc using libcyaml.
+ ******************************************************************************/
+/* Our CYAML config.
+ *
+ * If you don't want to change it between calls, make it const.
+ *
+ * Here we have a very basic config.
+ */
+cyaml_config_t cyaml_config = {
+    .log_fn = cyaml_log,            /* Use the default logging function. */
+    .mem_fn = cyaml_mem,            /* Use the default memory allocator. */
+    .log_level = CYAML_LOG_WARNING, /* Logging errors and warnings only. */
+    /*  .log_level = CYAML_LOG_DEBUG,   for logging debug. */
+};
+
+te_errno
+cyaml_process_errors(cyaml_err_t err)
+{
+    switch (err)
+    {
+        /**< Success. */
+        case CYAML_OK:
+            return 0;
+
+        /* Memory allocation failed. */
+        case CYAML_ERR_OOM:
+            ERROR("There is a CYAML error %s(%d)", cyaml_strerror(err), err);
+            return TE_ENOMEM;
```

Comment:
I suggest using `TE_FATAL_ERROR` as with other out-of-memory conditions.

Reply:
I've switched to `TE_FATAL_ERROR()` here. However I think that in others `conf_*` there are a lot of returns of `TE_ENOMEM`. It will be a little bit strange to eliminate it just in some cases. I'm sure that there should be a separate commit that eliminates `TE_ENOMEM` there.